{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1e4c04b-4403-40e4-ae00-20a702463dfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf <- read.table(\"Homo_sapiens.GRCh38.109.gtf\", sep=\"\\t\", header=T)\n",
    "gtf1 <- gtf[which(gtf$gene == \"gene\"),]\n",
    "#so basically the gene section has other stuff other than gene in it, extracting only the gene bits\n",
    "rm(gtf)\n",
    "gtf1_annotation <- gtf1[,9]\n",
    "# there was a set called genes in this table in the 9th column which is what we need\n",
    "gtf1_annotation <- gtf1_annotation[grep(\"gene name \", gtf1_annotation )]\n",
    "gtf1_annotation <- gtf1_annotation[grep(\"protein coding\", gtf1_annotation )]\n",
    "# weeding out the genes that are not named and that do not code for proteins\n",
    "gtf1_annotation_lst <- strsplit(gtf1_annotation, \"; \")\n",
    "gtf1_annotation_df <- as.data.frame(gtf1_annotation_lst)\n",
    "gtf1_annotation_df <- as.data.frame(t(gtf1_annotation_df))\n",
    "# t to put the different types of annotation at the top to use like columns\n",
    "# t turns the df into a matrix, however we want to keep using stuff like names and rownames so we redo as.data.frame\n",
    "# now putting only what we need from that so the id and the symbol columns\n",
    "gtf1_annotation_df <- gtf1_annotation_df[,c(1,3)] \n",
    "names(gtf1_annotation_df) <- c(\"ensemblid\",\"symbol\")\n",
    "# want to avoid having two ids matching with one symbol and viceversa, cause shit happens apparently\n",
    "# therefore we create a new rowname that unites the id to its symbol\n",
    "rownames(gtf1_annotation_df) <- paste(gtf1_annotation_df$ensemblid, gtf1_annotation_df$symbol, sep=\":\")\n",
    "#now we remove the redundancies like the \"gene id\" and \"gene name\" prefixes from the actual data we need\n",
    "gtf1_annotation_df$ensemblid <- gsub(\"gene_id \", \"\", gtf1_annotation_df$ensemblid)\n",
    "gtf1_annotation_df$symbol <- gsub(\"gene_name \", \"\", gtf1_annotation_df$symbol)\n",
    "rownames(gtf1_annotation_df) <- paste(gtf1_annotation_df$ensemblid, gtf1_annotation_df$symbol, sep=\":\")\n",
    "# on the lookout for duplicated symbols\n",
    "duplicated_symbols <- gtf1_annotation_df$symbol[which(duplicated(gtf1_annotation_df$symbol))]\n",
    "# selecting anything that is not duplicated with !\n",
    "gtf1_annotation_df <- gtf1_annotation_df$symbol[!duplicated(gtf1_annotation_df$symbol)]\n",
    "# we just remove the duplicates, cannot keep one of them because we can't know which is good\n",
    "# let's check for duplicated ensebl ids (prof checked it was 0 but let's make sure)\n",
    "gtf1_annotation_df <- gtf1_annotation_df$ensemblid[!duplicated(gtf1_annotation_df$ensemblid)]\n",
    "\n",
    "# annotating the gene count\n",
    "# count_table contains our aligned counts\n",
    "#first of all we filter our annotation for what is present in out counts\n",
    "gtf1_annotation_filtered <- gtf1_annotation_df[which(gtf1_annotation_df$ensemblid %in% rownames(count_table)),]\n",
    "# bopping and reversing it, we intercepted with the gtf but now we intercept with the count table\n",
    "# basically filter out whatever is in the count table that wasnt in the gtf, cause there may be some non protein coding\n",
    "count_table1 <- count_table[which(rownames(count_table) %in% gtf1_annotation_df$ensemblid),]\n",
    "dim(gtf1_annotation_df)\n",
    "dim(gtf1_annotation_filtered)\n",
    "dim(count_table)\n",
    "dim(count_table1)\n",
    "count_table <- count_table1\n",
    "# check if the annotation and the rownames are in the same identical order\n",
    "identical(rownames(count_table), gtf1_annotation_filtered$ensemblid)\n",
    "# ordering by ensemblid both of them\n",
    "gtf1_annotation_filtered <- gtf1_annotation_filtered[order(gtf1_annotation_filtered$ensemblid),]\n",
    "count_table <- count_table[order(rownames(count_table)),]\n",
    "identical(rownames(count_table), gtf1_annotation_filtered$ensemblid)\n",
    "rownames(count_table) <- gtf1_annotation_filtered$symbol\n",
    "\n",
    "# rownames = NA is used because the row names are shifted to the right of one column\n",
    "# so you just say that the first colname does not exist (so the rownames do not have a header)\n",
    "write.table(count_table, \"/sharedFolder/dataset/annotated_counts.tsv\", sep=\"\\t\", col.names = NA) \n",
    "\n",
    "# DIMENSIONALITY REDUCTION & KMEAN CLUSTERING\n",
    "library(umap)\n",
    "library(ggplot2)\n",
    "library(ClusterR)\n",
    "\n",
    "umap.out <- umap(t(log2_cpm_mat), n_epochs = 1000, min_dist = 0.01, n_neighbors = 4) \n",
    "f=data.frame(x=as.numeric(umap.out$layout[,1]), y=as.numeric(umap.out$layout[,2]))\n",
    "\n",
    "sp <- ggplot(f, aes(x=x, y=y)) + geom_point(pch=19,cex=0.3)\n",
    "km <- KMeans_arma(f, clusters = 5, n_iter = 25, seed_mode = \"random_subset\", verbose = T, centroids = NULL)\n",
    "pr <- predict_KMeans(f, km)\n",
    "pr <- as.factor(pr)\n",
    "# pr is a vector containing info of the various cells, we expect it to be sorted as the input file sp\n",
    "class(km) <- \"matrix\"\n",
    "sp <- ggplot(f, aes(x=x,y=y)) + geom.point(pch=19,cex=0.5, color=pr)\n",
    "sp\n",
    "f[1:3,1:2] \n",
    "# the row names define the single cell\n",
    "# str(umap.out) in distances the rows are the names of your library\n",
    "rownames(f) <- names(log2_cpm_mat) \n",
    "#it works cause they are in the same order\n",
    "rownames(f) <- paste(rownames(f), as.character(pr), sep=\":\")\n",
    "# rownames is the name of the cell, pr is the cluster they belong to\n",
    "names(log2_cpm_mat) <- rownames(f)\n",
    "# in grep the dollar indicates the end of the name, we start making the individual clusters now (make function)\n",
    "# in apply row is 1 column is 2, and then do the sum\n",
    "# the sum calculates how many counts are in each cell/replicate\n",
    "# it will sum the counts for each row for the cells that belong to cluster one\n",
    "# same process for the other clusters\n",
    "cls1 <- apply(log2_cpm_mat[,grep(\":1$\", names(log2_cpm_mat)), 1, sum)\n",
    "cls2 <- apply(log2_cpm_mat[,grep(\":2$\", names(log2_cpm_mat)), 1, sum)\n",
    "cls3 <- apply(log2_cpm_mat[,grep(\":3$\", names(log2_cpm_mat)), 1, sum)\n",
    "cls4 <- apply(log2_cpm_mat[,grep(\":4$\", names(log2_cpm_mat)), 1, sum)\n",
    "cls5 <- apply(log2_cpm_mat[,grep(\":5$\", names(log2_cpm_mat)), 1, sum)\n",
    "# now combine them in a dataframe\n",
    "cls <- data.frame(cls1, cls2, cls3, cls4, cls5)\n",
    "# so this will have the counts for each cluster\n",
    "# wait he is talking about converting to cpm here, I think I got lost.\n",
    "col_sum <- apply(cls, 2, sum)\n",
    "#rotate the matrix to apply the division to each row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17b67fcb-45a9-4f83-a2d4-3ba0fbc9669f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#the cls will be ranked on the expression of the various data after the division\n",
    "cls_rank <- t(cls)/col_sum\n",
    "cls_rank <- as.data.frame(t(cls_rank) * 1000000)\n",
    "cls_rank_log <- log2(cls_rank_log2 + 1)\n",
    "# the issue you cannot compare the two data (so the og log2cmp and the one you got here\n",
    "# cause they are in different scales but you can represent the data in a new way with their mean\n",
    "cls_rank_log_mean <- apply(cls_rank_log, 1, mean)\n",
    "cls_rank_log_normalised <- cls_rank_log - cls_rank_log_mean\n",
    "# it is the log2 ratio with respect to the mean of the row \n",
    "# you do the same for your log2cmp database\n",
    "log2_cpm_mat_mean <- apply(log2_cpm_mat, 1, mean)\n",
    "log2_cpm_mat_normalised <- log2_cpm_mat - log2_cpm_mat_mean\n",
    "# now combine all of this in one dataframe\n",
    "# log2=tian cls=cls\n",
    "cls_mat <- data.frame(log2_cpm_mat_normalised, cls_rank_log_normalised)\n",
    "# so now we can use the vector of the cell and the cluster to check if they resemble each other\n",
    "# in a hieracrchical clustering\n",
    "set.seed(111)\n",
    "cls_tian"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
