<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper function for circRNAs prediction using STARChip — wrapperSTARChip • docker4seq</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Wrapper function for circRNAs prediction using STARChip — wrapperSTARChip" />
<meta property="og:description" content="This function calls sequentially the docker containers for FASTQC, STAR, and STARChip to predict the list of circRNAs starting from the raw RNA-Seq reads" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper function for circRNAs prediction using STARChip</h1>
    
    <div class="hidden name"><code>wrapperSTARChip.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calls sequentially the docker containers for FASTQC, STAR, and STARChip to predict the list of circRNAs starting from the raw RNA-Seq reads</p>
    </div>

    <pre class="usage"><span class='fu'>wrapperSTARChip</span>(
  <span class='kw'>group</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"sudo"</span>, <span class='st'>"docker"</span>),
  <span class='no'>scratch.folder</span>,
  <span class='no'>genome.folder</span>,
  <span class='no'>samples.folder</span>,
  <span class='no'>threads</span>,
  <span class='no'>chimSegmentMin</span>,
  <span class='no'>chimJunctionOverhangMin</span>,
  <span class='no'>reads.cutoff</span>,
  <span class='no'>min.subject.limit</span>,
  <span class='no'>do.splice</span>,
  <span class='no'>cpm.cutoff</span>,
  <span class='no'>subjectCPM.cutoff</span>,
  <span class='no'>annotation</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>group, </th>
      <td><p>a character string. Two options: sudo or docker, depending to which group the user belongs</p></td>
    </tr>
    <tr>
      <th>scratch.folder, </th>
      <td><p>a character string indicating the scratch folder where docker container will be mounted</p></td>
    </tr>
    <tr>
      <th>genome.folder, </th>
      <td><p>a character string indicating the folder where the indexed reference genome for STAR is located.</p></td>
    </tr>
    <tr>
      <th>samples.folder, </th>
      <td><p>the folder where are located all the subfolders of the samples processed with starChimeric</p></td>
    </tr>
    <tr>
      <th>threads, </th>
      <td><p>a number indicating the number of cores to be used from the application</p></td>
    </tr>
    <tr>
      <th>chimSegmentMin, </th>
      <td><p>is a positive integer indicating the minimal length of the overlap of a read to the chimeric element</p></td>
    </tr>
    <tr>
      <th>chimJunctionOverhangMin, </th>
      <td><p>is a positive integer indicating the minimum overhang for a chimeric junction</p></td>
    </tr>
    <tr>
      <th>reads.cutoff, </th>
      <td><p>Integer. Minimum number of reads crossing the circRNA backsplice required.</p></td>
    </tr>
    <tr>
      <th>min.subject.limit, </th>
      <td><p>Integer. Minimum number of individuals with readsCutoff reads required to carry forward a circRNA for analysis</p></td>
    </tr>
    <tr>
      <th>do.splice, </th>
      <td><p>true false. The splices within the circRNA be detected and reported. Linear splices are searched within each circRNA in each individual. Any linear splice with &gt;= 60% of the read count of the cRNA is considered a splice within the circRNA. Two files are then created, .consensus with most common splice pattern, and .allvariants with all reported splice patterns.</p></td>
    </tr>
    <tr>
      <th>cpm.cutoff, </th>
      <td><p>Float. Reads counts are loaded into R and log2(CountsPerMillion) is calculated using the limma package. With cpmCutoff &gt; 0, circRNAs with log2(CPM) below this value will be filtered from this analysis</p></td>
    </tr>
    <tr>
      <th>subjectCPM.cutoff, </th>
      <td><p>Integer. See above. This value is the lower limit for number of individuals required to have the circRNAs expressed at a value higher than cpmCutoff.</p></td>
    </tr>
    <tr>
      <th>annotation, </th>
      <td><p>true/false. circRNAs are provided with gene annotations</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>1. Count matrices : raw cRNA backsplice counts: circRNA.cutoff[readthreshold]reads.[subjectthreshold]ind.countmatrix log2CPM of above: norm_log2_counts_circRNA.[readthreshold]reads.[subjectthreshold]ind.0cpm_0samples.txt Maximum Linear Splices at Circular Loci: rawdata/linear.[readthreshold]reads.[subjectthreshold]ind.sjmax 2. Info about each circRNA:  Consensus Information about Internal Splicing: Circs[reads].[subjects].spliced.consensus Complete Gene Annotation: circRNA.[readthreshold]reads.[subjectthreshold]ind.annotated Consise Gene Annotation + Splice Type:  circRNA.[readthreshold]reads.[subjectthreshold]ind.genes 3. Images: PCA plots: circRNA.[readthreshold]reads.[subjectthreshold]ind.0cpm_0samples_variance_PCA.pdf Heatmap: circRNA.[readthreshold]reads.[subjectthreshold]ind.heatmap.pdf</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {

    <span class='co'>#retrieve the example data</span>
    <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"wget https://github.com/carlo-deintinis/circhunter/archive/master.zip"</span>) <span class='co'>#retrieve the data of the indexed genome (chromosome 21 of hg38 human genome assembly)</span>
    <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"unzip master.zip"</span>)
    <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"unzip ./circhunter-master/CircHunter/data/hg38.chr21.fa.zip"</span>)
    <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR582/001/SRR5824251/SRR5824251_1.fastq.gz"</span>) <span class='co'>#retrieve the RNA-Seq data</span>
    <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR582/001/SRR5824251/SRR5824251_2.fastq.gz"</span>) <span class='co'>#retrieve the RNA-Seq data</span>

    <span class='co'>#running the wrapperSTARChip function</span>
<span class='fu'>wrapperSTARChip</span>(<span class='kw'>group</span> <span class='kw'>=</span> <span class='st'>"docker"</span>, <span class='kw'>scratch.folder</span><span class='kw'>=</span><span class='st'>"/data/scratch"</span>, <span class='kw'>genome.folder</span><span class='kw'>=</span><span class='st'>"./circhunter-master/CircHunter/data/"</span>, <span class='kw'>samples.folder</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span>(), <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>8</span>, <span class='kw'>chimSegmentMin</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>chimJunctionOverhangMin</span> <span class='kw'>=</span> <span class='fl'>15</span>, <span class='kw'>reads.cutoff</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>min.subject.limit</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>do.splice</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>cpm.cutoff</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>subjectCPM.cutoff</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Nicola Licheri, nicola [dot] licheri [at] unito [dot] it, University of Turin, Italy</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


