#' @title A function to associate CNV generated by oncSNP starting from log R ratio and B allel frequences, generated by SNP arrays, to gene symbols
#' @description This function executes a ubuntu docker that uses as input the cnvs files produced by oncoSNP and add CN and LOH (check oncoSNP readme for more info) fields (rank 5 only) to the genes characterized by having genomics cohordinates included in the CNV region
#' @param group, a character string. Two options: sudo or docker, depending to which group the user belongs
#' @param genome.folder, a character string indicating the path of the folder where ENSEMBL genome.gtf is located, IMPORTANT use the same genome assembly used to design the array.
#' @param data.folder, a character string indicating the folder where input data (.cnvs) generated with oncoSNP are located and where output will be written
#' @author Raffaele Calogero, raffaele.calogero[at]unito[dot]it, University of Torino, Italy
#'
#' @return a file for each .cnvs, e.g. annotation_XXXX.cnvs.txt
#' @examples
#' \dontrun{
#'     #running oncosnpAnnotation
#' system("wget http://130.192.119.59/public/test_oncosnp.zip")
#' system("unzip test_oncosnp.zip")
#' oncosnpAnnotation(group="docker", data.folder="./test_oncosnp/oncosnp_out", genome.folder="./test_oncosnp/hg19")
#' }
#'
#' @export
oncosnpAnnotation <- function(group=c("sudo","docker"), data.folder, genome.folder){


  #storing the position of the home folder
  home <- getwd()

  #running time 1
  ptm <- proc.time()
  #setting the data.folder as working folder
  if (!file.exists(data.folder)){
    cat(paste("\nIt seems that the ",data.folder, " folder does not exist\n"))
    return(2)
  }
  
  setwd(data.folder)
  #initialize status
  system("echo 0 > ExitStatusFile 2>&1")
  
  #testing if docker is running
  test <- dockerTest()
  if(!test){
    cat("\nERROR: Docker seems not to be installed in your system\n")
    system("echo 10 > ExitStatusFile 2>&1")
    setwd(home)
    return(10)
  }
  
  #check  if genome folder exist
  if (!file.exists(genome.folder)){
    cat(paste("\nIt seems that the ",genome.folder, " folder does not exist\n"))
    system("echo 3 > ExitStatusFile 2>&1")
    setwd(home)
    return(3)
  }
  #executing the docker job
    params <- paste("--cidfile ",data.folder,"/dockerID -v ",genome.folder,":/index -v ", data.folder, ":/data -d docker.io/repbioinfo/r340.2017.01 Rscript /bin/annotating_oncosnp.R", sep="")
    resultRun <- runDocker(group=group, params=params)

  #waiting for the end of the container work
  if(resultRun==0){
    cat("\nAnalysis is finished\n")
  }
  #running time 2
  ptm <- proc.time() - ptm
  dir <- dir(data.folder)
  dir <- dir[grep("run.info",dir)]
  if(length(dir)>0){
    con <- file("run.info", "r")
    tmp.run <- readLines(con)
    close(con)
    tmp.run[length(tmp.run)+1] <- paste("user run time mins ",ptm[1]/60, sep="")
    tmp.run[length(tmp.run)+1] <- paste("system run time mins ",ptm[2]/60, sep="")
    tmp.run[length(tmp.run)+1] <- paste("elapsed run time mins ",ptm[3]/60, sep="")
    writeLines(tmp.run,"run.info")
  }else{
    tmp.run <- NULL
    tmp.run[1] <- paste("run time mins ",ptm[1]/60, sep="")
    tmp.run[length(tmp.run)+1] <- paste("system run time mins ",ptm[2]/60, sep="")
    tmp.run[length(tmp.run)+1] <- paste("elapsed run time mins ",ptm[3]/60, sep="")

    writeLines(tmp.run,"run.info")
  }

  #saving log and removing docker container
  container.id <- readLines(paste(data.folder,"/dockerID", sep=""), warn = FALSE)
  system(paste("docker logs ", substr(container.id,1,12), " &> ",data.folder,"/", substr(container.id,1,12),".log", sep=""))
 # system(paste("docker rm ", container.id, sep=""))
  #removing temporary data
  system("rm -fR dockerID")
  system(paste("cp ",paste(path.package(package="docker4seq"),"containers/containers.txt",sep="/")," ",data.folder, sep=""))
  setwd(home)
}
