<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single Cell workflow • docker4seq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single Cell workflow">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/casc.html">Single Cell workflow</a>
    </li>
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Single Cell workflow</h1>
                        <h4 class="author">Raffaele A Calogero</h4>
            
      
      
      <div class="hidden name"><code>casc.Rmd</code></div>

    </div>

    
    
<div id="single-cell-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#single-cell-workflow" class="anchor"></a>Single Cell workflow</h2>
<ul>
<li>
<strong>inDropseq</strong>:
<ul>
<li>Creating a reference genome for inDrop V2:</li>
<li>From fastq to UMI counts using inDrop workflow: <strong>indropCounts</strong>
</li>
</ul>
</li>
<li>
<strong>Counts manipulation</strong>:
<ul>
<li>Removing non informative genes: <strong>allZeros</strong>
</li>
<li>Data normalization: <strong>scnorm</strong>
</li>
<li>Imputing dropouts: <strong>cascImpute</strong>
</li>
<li>Converting a count table in log10: <strong>counts2log</strong>
</li>
</ul>
</li>
<li>
<strong>Cell filtering</strong>:
<ul>
<li>Removing low quality cells: <strong>lorenzFilter</strong>
</li>
</ul>
</li>
</ul>
</div>
<div id="indrop-seq" class="section level2">
<h2 class="hasAnchor">
<a href="#indrop-seq" class="anchor"></a>inDrop seq</h2>
<p>The function indropCounts start for inDrop V2 library and convert fastq to UMI counts.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">indropCounts</span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, <span class="dt">fastq.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(),</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">        <span class="dt">index.folder=</span><span class="st">"/data/genomes/mm10indrop"</span>, <span class="dt">sample.name=</span><span class="st">"C2"</span>, <span class="dt">split.affixes=</span><span class="st">"S2_L001"</span>,</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">        <span class="dt">bowtie.index.prefix=</span><span class="st">"Mus_musculus.GRCm38.85.index"</span>, <span class="dt">M=</span><span class="dv">10</span>, <span class="dt">U=</span><span class="dv">2</span>, <span class="dt">D=</span><span class="dv">400</span>, <span class="dt">low.complexity.mask=</span><span class="st">"False"</span>)</a></code></pre></div>
</div>
<div id="counts-manipulation" class="section level2">
<h2 class="hasAnchor">
<a href="#counts-manipulation" class="anchor"></a>Counts manipulation</h2>
<div id="removing-non-informative-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#removing-non-informative-genes" class="anchor"></a>Removing non informative genes</h3>
<p>The function <strong>allZeros</strong> removed all genes that are without counts in all cells and plot the frequency distribution of genes with counts in the cells.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"gzip -d singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw">filterZeros</span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(),<span class="dt">counts.matrix=</span><span class="st">"singlecells_counts.txt"</span>)</a></code></pre></div>
</div>
<div id="data-normalization" class="section level3">
<h3 class="hasAnchor">
<a href="#data-normalization" class="anchor"></a>Data normalization</h3>
<p>SCnorm (as detailed in Bacher et al.) performs a quantile-regression based approach for robust normalization of single-cell RNA-seq data. SCnorm groups genes based on their count-depth relationship then applies a quantile regression to each group in order to estimate scaling factors which will remove the effect of sequencing depth from the counts.</p>
<div id="check-counts-depth-relationship" class="section level4">
<h4 class="hasAnchor">
<a href="#check-counts-depth-relationship" class="anchor"></a>Check counts-depth relationship</h4>
<p>Before normalizing using <strong>scnorm</strong>, it is advised to check the data count-depth relationship. <strong>checkCountDepth</strong> provides a wrapper, in docker4seq, for the checkCountDepth of the <a href="https://github.com/rhondabacher/SCnorm"><strong>SCnorm package</strong></a>, which estimates the count-depth relationship for all genes.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"gzip -d singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">conditions=<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span>,<span class="dv">288</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw">checkCountDepth</span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(), <span class="dt">counts.matrix=</span><span class="st">"singlecells_counts.txt"</span>, <span class="dt">conditions=</span>conditions, <span class="dt">outputName=</span><span class="st">"singlecells_counts"</span>, <span class="dt">nCores=</span><span class="dv">8</span>)</a></code></pre></div>
<p>The output is a PDF that provides a view of the counts distribution of the data.</p>
</div>
<div id="scnorm" class="section level4">
<h4 class="hasAnchor">
<a href="#scnorm" class="anchor"></a>scnorm</h4>
<p>the <strong>scnorm</strong> function execute SCnorm of the <a href="https://github.com/rhondabacher/SCnorm"><strong>SCnorm package</strong></a>, which normalizes across cells to remove the effect of sequencing depth on the counts and return the normalized expression count.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"gzip -d singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">conditions=<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span>,<span class="dv">288</span>)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw">scnorm</span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(),<span class="dt">counts.matrix=</span><span class="st">"singlecells_counts.txt"</span>, <span class="dt">conditions=</span>conditions,<span class="dt">outputName=</span><span class="st">"singlecells_counts"</span>, <span class="dt">nCores=</span><span class="dv">8</span>, <span class="dt">filtercellNum=</span><span class="dv">10</span>, <span class="dt">PropToUse=</span><span class="fl">0.1</span>)</a></code></pre></div>
<p>The output is a PDF that provides a view of effects of normalization, and Rda file containing the full output of <strong>SCnorm</strong> and a tab delimited file containing the normalized data.</p>
<p>Setting PropToUse to 0.1 was particularly useful with large UMI based datasets. Proportion of genes closest to the slope mode used for the group fitting. This number affects a lot speed.</p>
</div>
</div>
<div id="imputing-dropouts" class="section level3">
<h3 class="hasAnchor">
<a href="#imputing-dropouts" class="anchor"></a>Imputing dropouts</h3>
<p>the <strong>cascImpute</strong> function execute scImpute of the <a href="https://github.com/Vivianstats/scImpute"><strong>scImpute package</strong></a>, which impute the dropout values in scRNA-seq data.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"gzip -d singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="kw">cascImpute</span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(), <span class="dt">counts.matrix=</span><span class="st">"singlecells_counts.txt"</span>, <span class="dt">drop.thre=</span><span class="fl">0.5</span>, <span class="dt">cores=</span><span class="dv">8</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#Modifying drop.thre value A quick version of the imputing can be used to refine drop.thre values indicating refining=TRUE. It has to be done in the same folder where the frst run was done.</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="kw">cascImpute</span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(), <span class="dt">counts.matrix=</span><span class="st">"singlecells_counts.txt"</span>, <span class="dt">drop.thre=</span><span class="fl">0.3</span>, <span class="dt">cores=</span><span class="dv">8</span>, <span class="dt">refining=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<p>The output is a matrix file containing the imputed data.</p>
</div>
<div id="converting-a-count-table-in-log10" class="section level3">
<h3 class="hasAnchor">
<a href="#converting-a-count-table-in-log10" class="anchor"></a>Converting a count table in log10</h3>
<p>The function <strong>counts2log</strong> can convert a count table in a log10 values saved in a comma separated file which is the input of CASC.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"gzip -d singlecells_counts.txt.gz"</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="kw">counts2log</span>(<span class="dt">counts.matrix=</span><span class="st">"singlecells_counts.txt"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(), <span class="dt">log.base=</span><span class="dv">10</span>, <span class="dt">type=</span><span class="st">"txt"</span>)</a></code></pre></div>
</div>
</div>
<div id="cell-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-filtering" class="anchor"></a>Cell filtering</h2>
<div id="identifying-outlier-libraries" class="section level3">
<h3 class="hasAnchor">
<a href="#identifying-outlier-libraries" class="anchor"></a>Identifying outlier libraries</h3>
<p>To identify outlier libraries, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937196/">Diaz et al.2016</a> developed a strategy to estimate genes expressed at background levels in a given sample. Then samples whose background fraction is significantly larger than average is filtered out (Lorenz statistics). Furthermore, this statistics correlates with correlates (Pearson-correlation 0.7) with live-dead staining <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937196/">Diaz et al.2016</a>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">lorenzFilter</span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"path/of/scratch/folder"</span>,</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">           <span class="dt">data.folder=</span><span class="st">"path/of/data/folder"</span>,<span class="dt">matrixName=</span><span class="st">"matrixName"</span>,<span class="dt">p_value=</span><span class="fl">0.05</span>,<span class="dt">format=</span><span class="st">"txt"</span>,<span class="dt">separator=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</a></code></pre></div>
</div>
<div id="casc-classification-analysis-of-single-cell-sequencing-data" class="section level3">
<h3 class="hasAnchor">
<a href="#casc-classification-analysis-of-single-cell-sequencing-data" class="anchor"></a>CASC Classification Analysis of Single Cell Sequencing Data</h3>
<p>Genome-wide single-cell measurements such as transcriptome sequencing enable the characterization of cellular composition as well as functional variation in homogenic/heterogenic cell populations. An important step in the single-cell transcriptome analysis is to group cells that belong to the same sub-type based on gene expression patterns [Usoskin et al, Pollen et al, Kolodziejczyk et al]. Critical issues in cell clustering are:</p>
<ul>
<li><p>cluster stability</p></li>
<li><p>feature selection, i.e. the identification genes playing the major role in cluster formation.</p></li>
</ul>
<p>To address the above issues, we have developed CASC, a tool implemented in a docker container, that uses as core application to detect cell clusters the “kernel based similarity learning” [Wang et al] and allows:</p>
<ul>
<li><p>identification of the optimal number of clusters for cell partitioning using “silhouette method”.</p></li>
<li><p>The evaluation of clusters stability, measuring the permanence of a cell in a cluster upon random removal of subsets of cells.</p></li>
</ul>
<p>-Feature selection via “nearest shrunken centroid method” [Tibshirani et al], applied to the gene Index Of Dispersion [Diaz et al].</p>
<p>CASC was tested on previously published data sets [Usoskin et al, Pollen et al, Kolodziejczyk et al]. CASC feature selection procedure efficiently allows the identification of the subpopulation of genes playing the main role in discriminating between cell subpopulations.</p>
<div id="introduction" class="section level4">
<h4 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h4>
<p><strong>CASC</strong> is controlled by a set of functions embedded in <strong>kendomaniac/docker4seq</strong> package, which was developed to facilitate the use of computing demanding applications in the field of NGS data analysis.</p>
<p>This approach provides multiple advantages:</p>
<ul>
<li><p>user does not need to install all the software on its local server</p></li>
<li><p>results generated by different containers can be organized in pipelines</p></li>
<li><p>reproducible research is guarantee by the possibility of sharing the docker containers used for the analysis</p></li>
</ul>
</div>
<div id="requirements" class="section level4">
<h4 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h4>
<p>The minimal hardware requirements are a 4 core 64 bits linux computer, 32 Gb RAM, one SSD 250GB, with a folder with read/write permission for any users (chmod 777), and <a href="https://www.docker.com/">docker</a> installed.</p>
<p><strong>IMPORTANT</strong> The first time <em>docker4seq</em> is installed the <strong>downloadContainers</strong> needs to be executed to download to the local repository the containers that are needed for the use of <em>docker4seq</em></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="../reference/downloadContainers.html">downloadContainers</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>)</a></code></pre></div>
</div>
<div id="defining-the-optimal-number-of-clusters" class="section level4">
<h4 class="hasAnchor">
<a href="#defining-the-optimal-number-of-clusters" class="anchor"></a>Defining the optimal number of clusters</h4>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/log10_singlecells_counts.csv.gz"</span>)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"gzip -d log10_singlecells_counts.csv.gz"</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="kw">cascKoptimization</span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">scratch.folder=</span><span class="st">"/Users/raffaelecalogero/Desktop/scratch"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(),</a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="dt">counts.matrix=</span><span class="st">"log10_singlecells_counts.csv"</span>, <span class="dt">permutations=</span><span class="dv">20</span>, <span class="dt">blocks.permutations=</span><span class="dv">2</span>, <span class="dt">core=</span><span class="dv">0</span>, <span class="dt">bootstrap.fraction=</span><span class="dv">10</span>, <span class="dt">k.min=</span><span class="dv">2</span>, <span class="dt">k.max=</span><span class="dv">4</span>, <span class="dt">totIdentity=</span><span class="dv">80</span>, <span class="dt">clusterIdentity=</span><span class="dv">80</span>)</a></code></pre></div>
<p>The function <strong>cascKoptimization</strong> run SIMLR [Wang et al] using a range of clusters and produces as output two violin plots:</p>
<ul>
<li><p>silhouette.pdf: Silhouette refers to a method of interpretation and validation of consistency within clusters of data. The technique provides a succinct graphical representation of how well each object lies within its cluster. Here instead of using the average silhouette we represent the silhouette values distribution using a violin plot. Thus, silhouette distributions being skewed to the positive values and with short negative tail is representative of a consistent cluster. In the example below, fig. A, it seems that 4 clusters are the most consistent by silhouette analysis.</p></li>
<li><p>cell.stability.pdf: Cell stability plot represent the distribution of the fraction of times, given a N number of permutations, that the cell are stabily localized in a cluster. The example below, fig. B, 3 clusters are characterized by an higher cell stability with respect to 4 clusters.</p></li>
</ul>
<p>Taking in account the two plots and the clusters structure observed for 3 and 4 clusters, fig C and D, it is clear that 4 clusters provide a better separation between cells but keeping very high the cell permanence in a cluster.</p>
<div class="figure">
<img src="../inst/img/silhouette.stability.jpg" alt="cluster decision"><p class="caption">cluster decision</p>
</div>
<p>The function <strong>cascOutputReformat</strong> use SilhouetteParameters.csv (the file containing the cell silhouette scores), mainVector.csv (the file that associates each cell to a specific SIMLR cluster), scoreVector.csv (the file associating each cell to a specific cell stability score), dataPlot.csv (containing SIMLR component 1 and 2 cohordinates) to generate a summary file called <strong>summary_table.csv</strong>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"> <span class="co">#downloading fastq files</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">"wget http://130.192.119.59/public/example.zip"</span>)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span>(<span class="st">"example.zip"</span>)</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">"./example"</span>)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="kw">cascOutputReformat</span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>())</a></code></pre></div>
<div class="figure">
<img src="../inst/img/summary_table.jpeg" alt="summary_table"><p class="caption">summary_table</p>
</div>
</div>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>Usoskin et al. Nat. Neurosci. 2014, 18:145–153</p>
<p>Pollen et al. Nat. Biotechnol. 2014, 32:1–37</p>
<p>Kolodziejczyk et al. Cell Stem Cell 2015, 17:471–485</p>
<p>Wang et al. Nat Methods. 2017 14:414-416</p>
<p>Tibshirani et al. PNAS 2002, 99:6567-6572</p>
<p>Diaz et al. Bioinformatics 2016, 32: 2219-2220</p>
<p>Bacher et al. Nat. Methods 2017, 14:584–586</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#single-cell-workflow">Single Cell workflow</a></li>
      <li><a href="#indrop-seq">inDrop seq</a></li>
      <li><a href="#counts-manipulation">Counts manipulation</a></li>
      <li><a href="#cell-filtering">Cell filtering</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
